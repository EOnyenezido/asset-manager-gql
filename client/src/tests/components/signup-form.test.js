import React from 'react';
import { create } from "react-test-renderer";
import SignUpForm, { REGISTER_USER } from '../../components/signup-form';
import { MockedProvider } from 'react-apollo/test-utils';
import wait from 'waait';

// render
it('given no props, renders without crashing', () => {
  create(
    <MockedProvider mocks={[]}>
      <SignUpForm />
    </MockedProvider>);
});

// snapshot
describe('Snapshot testing', () => {
  it('matches the snapshot', () => {
    const component = create(
      <MockedProvider mocks={[]}>
      <SignUpForm />
    </MockedProvider>);
    expect(component.toJSON()).toMatchSnapshot();
  });
});

// registration function
describe('Registration function', () => {
  // success
  it('registers a user with accurate details', async () => {
    const response = { emailAddress: 'foo@bar.com', firstName: 'Foo', lastName: 'Bar', id: 1, token: 'thisisasupersecurehash' };
    const mocks = [
      {
        request: {
          query: REGISTER_USER,
          variables: {
            emailAddress: '',
            password: '',
            firstName: '',
            lastName: '',
            phoneNumber: ''
          },
        },
        result: { data: { register: response } },
      },
    ];
    const component = create(
      <MockedProvider mocks={mocks} addTypename={false}>
        <SignUpForm />
      </MockedProvider>);
    const submitForm = component.root.findByType('form');
    submitForm.props.onSubmit();
    await wait(10); // wait for simulated delay for UI
    const alert = submitForm.findByProps({role: "alert"});
    
    // no errors or alerts
    expect(alert.props.hidden).toBeTruthy();
  });

  // error
  it('displays an error for a registration with invalid details', async () => {
    const mocks = [
      {
        request: {
          query: REGISTER_USER,
          variables: {
            emailAddress: '',
            password: '',
            firstName: '',
            lastName: '',
            phoneNumber: ''
          },
        },
        result: { errors: [{ message: 'Field "password" must contain at least 10 characters!' }] }, // server says registration is invalid
      },
    ];
    const component = create(
      <MockedProvider mocks={mocks} addTypename={false}>
        <SignUpForm />
      </MockedProvider>);
    const submitForm = component.root.findByType('form');
    submitForm.props.onSubmit();
    await wait(10);
    const alert = submitForm.findByProps({role: "alert"});
    
    // alert message is shown and contains the message from the error thrown
    expect(alert.props.hidden).toBeFalsy();
    expect(alert.children[1].children).toContain('Field "password" must contain at least 10 characters!');
  });

  it('displays an error when connectivity cound not be established', async () => {
    const mocks = [
      {
        request: {
          query: REGISTER_USER,
          variables: {
            emailAddress: '',
            password: '',
            firstName: '',
            lastName: '',
            phoneNumber: ''
          },
        },
        error: new Error('Uh oh, cannot seem to connect to server'), // no connection
      },
    ];
    const component = create(
      <MockedProvider mocks={mocks} addTypename={false}>
        <SignUpForm />
      </MockedProvider>);
    const submitForm = component.root.findByType('form');
    submitForm.props.onSubmit();
    await wait(10);
    const alert = submitForm.findByProps({role: "alert"});
    
    // alert message is shown and contains the correct hardcoded message
    expect(alert.props.hidden).toBeFalsy();
    expect(alert.children[1].children).toContain('Unable to connect. Please check your connection and try again.');
  });
});