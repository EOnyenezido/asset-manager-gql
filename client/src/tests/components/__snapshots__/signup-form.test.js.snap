// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot testing matches the snapshot 1`] = `
<div
  className="kt-login__signup undefined"
>
  <div
    className="kt-login__head"
  >
    <h3
      className="kt-login__title"
    >
      Request Access
    </h3>
    <div
      className="kt-login__desc"
    >
      Enter your details to request access:
    </div>
  </div>
  <form
    action=""
    className="kt-form"
    onSubmit={[Function]}
  >
    <div
      className="kt-alert kt-alert--outline alert alert-danger alert-dismissible animated fadeIn"
      hidden={true}
      role="alert"
    >
      <button
        aria-label="Close"
        className="close"
        data-dismiss="alert"
        type="button"
      />
      <span>
        
      </span>
    </div>
    <div
      className="input-group row"
    >
      <input
        className="form-control mr-2"
        name="firstName"
        onChange={[Function]}
        placeholder="First Name"
        type="text"
      />
      <input
        className="form-control ml-2"
        name="lastName"
        onChange={[Function]}
        placeholder="Last Name"
        type="text"
      />
    </div>
    <div
      className="input-group"
    >
      <input
        autoComplete="off"
        className="form-control"
        name="emailAddress"
        onChange={[Function]}
        placeholder="Email"
        type="text"
      />
    </div>
    <div
      className="input-group"
    >
      <input
        className="form-control"
        name="phoneNumber"
        onChange={[Function]}
        placeholder="Phone Number"
        type="text"
      />
    </div>
    <div
      className="input-group"
    >
      <input
        className="form-control mr-2"
        name="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
      />
      <input
        className="form-control ml-2"
        name="rpassword"
        onChange={[Function]}
        placeholder="Confirm Password"
        type="password"
      />
    </div>
    <div
      className="row kt-login__extra"
    >
      <div
        className="col kt-align-left"
      >
        <label
          className="kt-checkbox"
        >
          <input
            name="agree"
            type="checkbox"
          />
          I Agree the 
          <a
            className="kt-link kt-login__link kt-font-bold"
            href="/"
          >
            terms and conditions
          </a>
          .
          <span />
        </label>
        <span
          className="form-text text-muted"
        />
      </div>
    </div>
    <div
      className="kt-login__actions"
    >
      <button
        className="btn btn-brand btn-elevate kt-login__btn-primary "
        disabled={false}
        id="kt_login_signup_submit"
      >
        Request
      </button>
       
       
      <button
        className="btn btn-light btn-elevate kt-login__btn-secondary"
        id="kt_login_signup_cancel"
        onClick={[Function]}
      >
        Cancel
      </button>
    </div>
  </form>
</div>
`;
